src/librustc_typeck/check/coercion.rs

```diff
@@ -121,39 +131,107 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
-                    self.coerce_unsafe_ptr(a, b, mt_b.mutbl)
-                });
+                return self.coerce_unsafe_ptr(a, b, mt_b.mutbl);
             }
 
             ty::TyRef(_, mt_b) => {
-                return self.unpack_actual_value(a, |a| {
-                    self.coerce_borrowed_pointer(expr_a, a, b, mt_b.mutbl)
-                });
+                return self.coerce_borrowed_pointer(expr_a, a, b, mt_b.mutbl);
             }
 
             _ => {}
         }
 
-        self.unpack_actual_value(a, |a| {
-            match a.sty {
-                ty::TyBareFn(Some(_), a_f) => {
-                    // Function items are coercible to any closure
-                    // type; function pointers are not (that would
-                    // require double indirection).
-                    self.coerce_from_fn_item(a, a_f, b)
-                }
-                ty::TyBareFn(None, a_f) => {
-                    // We permit coercion of fn pointers to drop the
-                    // unsafe qualifier.
-                    self.coerce_from_fn_pointer(a, a_f, b)
-                }
-                _ => {
-                    // Otherwise, just use subtyping rules.
-                    self.subtype(a, b)
+        let b = self.fcx.infcx().shallow_resolve(b);
+        match a.sty {
+            ty::TyBareFn(Some(_), a_f) => {
+                // Function items are coercible to any closure
+                // type; function pointers are not (that would
+                // require double indirection).
+                return self.coerce_from_fn_item(a, a_f, b);
+            }
+            ty::TyBareFn(None, a_f) => {
+                // We permit coercion of fn pointers to drop the
+                // unsafe qualifier.
+                return self.coerce_from_fn_pointer(a, a_f, b);
+            }
+            _ => {}
+        }
+
+        // Attempt to generalize the expected type in hopes of an unsizing
+        // coercion, where an intermediary stop-gap is usually necessary.
+        // This is the case with trait method calls where the returned type
+        // was not inferred, e.g. `Make::make(x: T): Box<Trait>`, if `Make`
+        // has many implementations. Unsizing coercion will be ambiguous
+        // and subtyping would result in a selection failure, if `Box<Trait>`
+        // does not implement `Make`, but `Box<T>` does. The stop-gap fix
+        // is `Make::make(x: T): Box<T>: Box<Trait>`.
+        // In that same case, the following generalization will attempt to
+        // apply `Box<_>` to the otherwise unconstrained `Make::make` return
+        // type and trigger selection, hoping to get the unambiguous source
+        // type `Box<T>` for the coercion to `Box<Trait>`.
+        // Subtyping rules are used if the generalization and second attempt
+        // at coercions through unsizing do not apply.
+
+        // The first unsizing coercion must have failed due to ambiguity.
```

Comment:
The "must" refers to a condition for trying the generalization.
If the first unsizing coercion did not "fail due to ambiguity", then it "failed" due to certain inapplicability, so there is no point in trying harder.
The use of "fail" here was meant to be about the fact that the compiler didn't go through with the unsizing coercion at the start of the process.